src/crawlee/storage_clients/_file_system/_dataset_client.py

```diff
@@ -0,0 +1,483 @@
+from crawlee.storage_clients._base import DatasetClient
+from crawlee.storage_clients.models import DatasetItemsListPage, DatasetMetadata
+
+if TYPE_CHECKING:
+    from collections.abc import AsyncIterator
+
+    from crawlee.configuration import Configuration
+
+logger = getLogger(__name__)
+
+
+class FileSystemDatasetClient(DatasetClient):
+    """File system implementation of the dataset client.
+
+    This client persists dataset items to the file system as individual JSON files within a structured
+    directory hierarchy following the pattern:
+
+    ```
+    {STORAGE_DIR}/datasets/{DATASET_ID}/{ITEM_ID}.json
+    ```
+
+    Each item is stored as a separate file, which allows for durability and the ability to
+    recover after process termination. Dataset operations like filtering, sorting, and pagination are
+    implemented by processing the stored files according to the requested parameters.
+
+    This implementation is ideal for long-running crawlers where data persistence is important,
+    and for development environments where you want to easily inspect the collected data between runs.
+    """
+
+    _STORAGE_SUBDIR = 'datasets'
+    """The name of the subdirectory where datasets are stored."""
+
+    _STORAGE_SUBSUBDIR_DEFAULT = 'default'
+    """The name of the subdirectory for the default dataset."""
+
+    _ITEM_FILENAME_DIGITS = 9
+    """Number of digits used for the dataset item file names (e.g., 000000019.json)."""
+
+    def __init__(
+        self,
+        *,
+        id: str,
+        name: str | None,
+        created_at: datetime,
+        accessed_at: datetime,
+        modified_at: datetime,
+        item_count: int,
+        storage_dir: Path,
+    ) -> None:
+        """Initialize a new instance.
+
+        Preferably use the `FileSystemDatasetClient.open` class method to create a new instance.
+        """
+        self._metadata = DatasetMetadata(
+            id=id,
+            name=name,
+            created_at=created_at,
+            accessed_at=accessed_at,
+            modified_at=modified_at,
+            item_count=item_count,
+        )
+
+        self._storage_dir = storage_dir
+
+        # Internal attributes
+        self._lock = asyncio.Lock()
+        """A lock to ensure that only one operation is performed at a time."""
+
+    @property
+    @override
+    def metadata(self) -> DatasetMetadata:
+        return self._metadata
+
+    @property
+    def path_to_dataset(self) -> Path:
+        """The full path to the dataset directory."""
+        if self.metadata.name is None:
+            return self._storage_dir / self._STORAGE_SUBDIR / self._STORAGE_SUBSUBDIR_DEFAULT
+
+        return self._storage_dir / self._STORAGE_SUBDIR / self.metadata.name
+
+    @property
+    def path_to_metadata(self) -> Path:
+        """The full path to the dataset metadata file."""
+        return self.path_to_dataset / METADATA_FILENAME
+
+    @override
+    @classmethod
+    async def open(
```

Comment:
It looks like there is a lot of duplication between `FileSystemDatasetClient.open` and its counterparts for kv-stores and RQs. Is it worth refactoring or not really?

Reply:
That's all I needed to hear, let's keep it this way :slightly_smiling_face: